genesyscloud/util_files.go

```diff
@@ -2,119 +2,113 @@ package genesyscloud
 
 import (
 	"bytes"
-	"crypto/sha256"
-	"encoding/hex"
 	"fmt"
 	"io"
 	"io/ioutil"
+	"log"
 	"net/http"
 	"net/url"
 	"os"
 	"strings"
 )
 
-func downloadOrOpenFile(path string) (io.Reader, *os.File, error) {
-	var reader io.Reader
-	var file *os.File
-
-	_, err := os.Stat(path)
-	if err != nil {
-		_, err = url.ParseRequestURI(path)
-		if err == nil {
-			resp, err := http.Get(path)
-			if err != nil {
-				return nil, nil, err
-			}
-			if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
-				return nil, nil, fmt.Errorf("HTTP Error downloading file: %v", resp.StatusCode)
-			}
-			reader = resp.Body
-		} else {
-			return nil, nil, fmt.Errorf("Invalid file path or URL: %v", path)
-		}
-	} else {
-		file, err = os.Open(path)
-		if err != nil {
-			return nil, nil, err
-		}
-		reader = file
-	}
-
-	return reader, file, nil
+type S3Uploader struct {
+	reader        io.Reader
+	substitutions map[string]interface{}
+	headers       map[string]string
+	presignedUrl  string
+	client        http.Client
 }
 
-// Hash file content, used in stateFunc for "filepath" type attributes
-func hashFileContent(path string) string {
-	reader, file, err := downloadOrOpenFile(path)
-	if err != nil {
-		return err.Error()
-	}
-	if file != nil {
-		defer file.Close()
+func NewS3Uploader(reader io.Reader, substitutions map[string]interface{}, headers map[string]string, presignedUrl string) *S3Uploader {
+	c := &http.Client{}
+	s3Uploader := &S3Uploader{
+		reader:        reader,
+		substitutions: substitutions,
+		headers:       headers,
+		presignedUrl:  presignedUrl,
+		client:        *c,
 	}
 
-	hash := sha256.New()
-	if file == nil {
-		if _, err := io.Copy(hash, reader); err != nil {
-			return err.Error()
-		}
-	} else {
-		if _, err := io.Copy(hash, file); err != nil {
-			return err.Error()
-		}
-	}
-
-	return hex.EncodeToString(hash.Sum(nil))
+	log.Printf("%#v\n", s3Uploader)
+	return s3Uploader
 }
 
-// Read and upload input file path to S3 pre-signed URL
-func prepareAndUploadFile(filename string, substitutions map[string]interface{}, headers map[string]string, presignedUrl string) ([]byte, error) {
-	bodyBuf := &bytes.Buffer{}
-
-	reader, file, err := downloadOrOpenFile(filename)
-	if err != nil {
-		return nil, err
-	}
-	if file != nil {
-		defer file.Close()
-	}
-
-	_, err = io.Copy(bodyBuf, reader)
-	if err != nil {
-		return nil, fmt.Errorf("Failed to copy file content to the handler. Error: %s ", err)
-	}
-
+func (s *S3Uploader) substituteValues(bodyBuf *bytes.Buffer) {
```

Comment:
Broke this out of the upload method and made this its own method.  This code can be a standalone method and also makes it easier to unit test.